sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -620,6 +620,12 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     // We can return what the children return. Same thing should happen in the codegen path.
     if (DataType.equalsStructurally(from, to)) {
       identity
+    } else if (from == NullType) {
+      // According to `canCast`, NullType can be casted to any type.
+      // For primitive types, we don't reach here because the guard of `nullSafeEval`.
+      // But for nested types like struct, we might reach here for nested null type field.
+      // We won't call the returned function actually, but returns a placeholder.
+      _ => throw new SparkException(s"should not directly cast from NullType to $to.")
     } else {
       to match {
         case dt if dt == from => identity[Any]
```

Comment:
Instead of adding `else if (from == NullType)` to add a place holder, how about the following?
```scala
- case dt if dt == from => identity[Any]
+ case dt if dt == from || from == NullType => identity[Any]
```

Reply:
When `from` is `NullType`, the returned function won't be called, actually, but just a placeholder.

Instead of `identity`, throwing an exception sounds better, so it prevents accidentally calling it.

Reply:
Got it. Although it's not called, yes. That would be safer.